src/Sarif/Writers/SarifLogger.cs

```diff
@@ -40,17 +40,25 @@ public class SarifLogger : IDisposable, IAnalysisLogger
 
             if (analysisTargets != null)
             {
-                run.Files = new Dictionary<string, FileData>();
+                run.Files = new List<FileData>();
 
                 foreach (string target in analysisTargets)
                 {
-                    string fileDataKey = UriHelper.MakeValidUri(target);
+                    Uri uri = new Uri(UriHelper.MakeValidUri(target));
 
                     var fileData = FileData.Create(
                         new Uri(target, UriKind.RelativeOrAbsolute),
                         dataToInsert);
 
-                    run.Files[fileDataKey] = fileData;
+                    var fileLocation = new FileLocation
+                    {
+                        Uri = uri
+                    };
+
+                    fileData.FileLocation = fileLocation;
+
+                    // This call will insert the file object into run.Files if not already present
+                    fileData.FileLocation.FileIndex = run.GetFileIndex(fileData.FileLocation, addToFilesTableIfNotPresent: true);
```

Comment:
>addToFilesTableIfNotPresent [](start = 94, length = 27)

Could we avoid this by starting out with a hash set? Well, I suppose that won't work because our ValueComparer includes `fileIndex`...

Hey! I could add a new type of code gen hint that would let you omit properties from the value comparer, what do you think?

Reply:
I don't see how your comment is relevant here. 'addToFilesTableIfNotPresent' simply tells the underlying helper whether we are simply querying for the existing index or if it should be added. That underlying code maintains a dictionary of key to the relevant index. So not sure how a hash set helps.

However, yes, the general improvement you're pointing to is quite interesting: how do we capture relevant identity semantics in our complex format in a way that's broadly helpful. What we need here is a special value comparer that's used strictly for the files table. The code gen could then usefully emit a backing Dictionary<FileLocation, int> item that we'd consult in order to manage/populate this thing. We could perform similar tricks for rules and logical locations table (that latter table should have a unique array as well, it's a bug IMO that it isn't expressed this way today)

---
In reply to: [244521472](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244521472) [](ancestors = 244521472)